keps/sig-api-machinery/3962-mutating-admission-policies/README.md

```diff
@@ -0,0 +1,1514 @@
+
+With mutating admission policies added, the the mutating admission plugin order will become:
+- Mutating admission controllers(e.g. DefaultIngressClass, DefaultStorageClass, etc)
+- Mutating admission policies (introduced within this enhancement and the order will be discussed below)
+- Mutating admission webhooks (ordered lexicographically by webhook name)
+
+To allow mutating admission plugins to observe changes made by other plugins, built-in mutating admission plugins are re-run if a mutating webhook modifies an object, 
+same will apply with mutating policy. The mutating policies are rerun if a mutating webhook or mutation policy modifies an object.
+
+For the running order within mutating admission policies, there are a couple options proposed:
+- option 1(suggested by @deads2k): ordered randomly but keep the same random order while reinvocation.
+  - Pros: 
+    - Encourage user to write order-independent mutations
+  - Cons: 
+    - The final state of request is not deterministic
+    - The mutation should not have dependencies in between
+- option 2: the lexicographical ordering of the resource names
+  - Pros:
+    - Align with the behavior with mutating webhook
+  - Cons:
+    - User has to be mindful on the order if there is dependency existing
+    - User has a hacky way to enforce the order
+
+Considering it would be easier to go with random order and then switch to a particular order, 
+
+Notes: If the mutations run in random order, a concern would be if people didn't write idempotence mutations,
+the result might be different between two admission request. Please refer to Safety session for ways to check idempotence.
+
+#### Metrics
+Goals:
+- Parity with validating admission policy metrics 
+  - Should include counter of deny, success violations 
+  - Label by {policy, policy binding, mutation expression} identifiers
+- Counters for number of policy definitions and policy bindings in cluster 
+  - Label by state (active vs. error), enforcement action (deny, warn)
+- Counters for Variable Composition 
+  - Should include a counter of variable resolutions to measure time saved. 
+  - Label by policy identifier
+
+
+### Phase 2
+
+All these capabilities are required and should be discussed thoughtfully before Beta, but will not be implemented in the first alpha release of this enhancement due to the size and complexity of this enhancement.
+
+#### Construct Type Enforcement
+
+The type alias Object and its descendants, in Phase 2, are now real types that derive from the resolved OpenAPI schemas. 
+If any type violations happen in the constructed object, the CEL checker will raise the errors before the expressions evaluate.
+
+For bigger schemas, the construction of CEL types can be expensive. It is recorded to take ~100ms to resolve and parse apps/v1.Deployment. 
+Optimizations like caching or lazy sub-schema resolution can be candidates of beta/GA graduation criteria.
+
+#### Unsetting values
```

Comment:
This section seems fairly complex, have we thought about how this will end up being documented for end users of the feature? I'm wondering if there's a way to extend the CEL functionality to give what you have in this section, but with a more intuitive syntax

Reply:
Thanks! Yes this is for addressing the field removal use cases which is not covered by structural merge very well. Ideally it should be very similar with how to modify the field but with optional.none() to indicate the deletion. A helper alike function was considered and we'll see how confuse it is in alpha before adding helper func. And YES we will need to carefully document it with tons of examples. :)